File: scripts/fetch_all_data.py
import requests
import json
import os
import time

BASE_URL = "https://solve.ivy.homes"
API_KEY = "IVY26-4C3EAEB6A76C"

DATA_DIR = os.path.join(os.path.dirname(__file__), "..", "data")
os.makedirs(DATA_DIR, exist_ok=True)

def log(msg):
    print(f"[FETCH] {msg}")

# 1. Health check
def fetch_health():
    log("Checking /health...")
    try:
        r = requests.get(f"{BASE_URL}/health")
        log(f"Health response: status={r.status_code}, body={r.text}")
        with open(os.path.join(DATA_DIR, "health.json"), "w", encoding="utf-8") as f:
            f.write(r.text)
    except Exception as e:
        log(f"Health error: {e}")

# 2. Auth login test
def fetch_auth():
    log("Testing /auth/login...")
    payload = {
        "email": "demo1@ivy.homes",
        "password": "117e45bfc1"
    }
    r = requests.post(f"{BASE_URL}/auth/login", json=payload)
    log(f"Auth login status: {r.status_code}, body={r.text}")
    if r.status_code == 200:
        token_data = r.json()
        with open(os.path.join(DATA_DIR, "auth_token.json"), "w", encoding="utf-8") as f:
            json.dump(token_data, f, indent=2)
        return token_data.get("token")
    return None

# 3. Paginated generic fetcher
def fetch_paginated_endpoint(path, filename, limit=200):
    log(f"Fetching endpoint {path} with limit={limit}...")
    page = 1
    all_records = []
    first_response = None
    
    while True:
        url = f"{BASE_URL}{path}?api_key={API_KEY}&page={page}&limit={limit}"
        r = requests.get(url)
        if r.status_code != 200:
            log(f"Error fetching page {page}: {r.status_code} {r.text}")
            break
        
        data = r.json()
        if first_response is None:
            first_response = data
            
        results = data.get("results", [])
        if not results and isinstance(data, list):
            results = data
            
        all_records.extend(results)
        
        total = data.get("total")
        page_size = data.get("page_size", limit)
        
        log(f"Fetched page {page}: got {len(results)} items (Total reported: {total})")
        
        if not results:
            break
        if total is not None and len(all_records) >= total:
            break
        if len(results) < limit:
            break
            
        page += 1
        time.sleep(0.05) # respect rate limit

    log(f"Total fetched for {path}: {len(all_records)}")
    output_path = os.path.join(DATA_DIR, filename)
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump({
            "first_response_meta": first_response,
            "total_records_fetched": len(all_records),
            "records": all_records
        }, f, indent=2)
    return all_records

# 4. Fetch Analytics Summary
def fetch_analytics():
    log("Fetching /v1/analytics/summary...")
    url = f"{BASE_URL}/v1/analytics/summary?api_key={API_KEY}"
    r = requests.get(url)
    log(f"Analytics status: {r.status_code}")
    if r.status_code == 200:
        with open(os.path.join(DATA_DIR, "analytics.json"), "w", encoding="utf-8") as f:
            json.dump(r.json(), f, indent=2)

if __name__ == "__main__":
    fetch_health()
    token = fetch_auth()
    
    log("Fetching listings...")
    listings = fetch_paginated_endpoint("/v1/listings", "listings.json")
    
    log("Fetching rentals...")
    rentals = fetch_paginated_endpoint("/v1/rentals", "rentals.json")
    
    log("Fetching projects...")
    projects = fetch_paginated_endpoint("/v1/projects", "projects.json")
    
    fetch_analytics()
    log("Data collection phase complete!")

File: scripts/fetch_all_data.js
const fs = require('fs');
const path = require('path');

const BASE_URL = 'https://solve.ivy.homes';
const API_KEY = 'IVY26-4C3EAEB6A76C';

const DATA_DIR = path.join(__dirname, '..', 'data');
if (!fs.existsSync(DATA_DIR)) {
  fs.mkdirSync(DATA_DIR, { recursive: true });
}

function log(msg) {
  console.log(`[FETCH] ${msg}`);
}

async function fetchHealth() {
  log('Checking /health...');
  try {
    const res = await fetch(`${BASE_URL}/health`, {
      headers: { 'X-API-Key': API_KEY }
    });
    const text = await res.text();
    log(`Health response status: ${res.status}, body: ${text}`);
    fs.writeFileSync(path.join(DATA_DIR, 'health.json'), text, 'utf-8');
  } catch (err) {
    log(`Health error: ${err.message}`);
  }
}

async function fetchAuth() {
  log('Testing /auth/login...');
  try {
    const res = await fetch(`${BASE_URL}/auth/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-API-Key': API_KEY
      },
      body: JSON.stringify({
        email: 'demo1@ivy.homes',
        password: '117e45bfc1'
      })
    });
    const text = await res.text();
    log(`Auth response status: ${res.status}, body: ${text}`);
    if (res.ok) {
      const data = JSON.parse(text);
      fs.writeFileSync(path.join(DATA_DIR, 'auth_token.json'), JSON.stringify(data, null, 2));
      return data.access_token || data.token;
    }
  } catch (err) {
    log(`Auth error: ${err.message}`);
  }
  return null;
}

async function fetchAllPages(endpointPath, fileName, token) {
  log(`Starting full pagination fetch for ${endpointPath}...`);
  let page = 1;
  const allRecords = [];
  let firstResponse = null;

  const headers = {
    'X-API-Key': API_KEY,
    'Authorization': `Bearer ${token}`
  };

  while (true) {
    // Request with limit 200 or default limit to see max returned
    const url = `${BASE_URL}${endpointPath}?page=${page}&limit=200`;
    const res = await fetch(url, { headers });
    if (!res.ok) {
      log(`Error fetching page ${page}: status ${res.status}`);
      const errText = await res.text();
      log(`Error detail: ${errText}`);
      break;
    }
    const data = await res.json();
    if (!firstResponse) {
      firstResponse = data;
    }

    const results = data.results || (Array.isArray(data) ? data : []);
    if (!results || results.length === 0) {
      log(`Page ${page} returned 0 results. Stopping.`);
      break;
    }

    allRecords.push(...results);
    const total = data.total;
    log(`Page ${page}: got ${results.length} items (Accumulated: ${allRecords.length} / Total: ${total})`);

    if (total !== undefined && allRecords.length >= total) {
      log(`Reached total of ${total} records.`);
      break;
    }

    page++;
    await new Promise((r) => setTimeout(r, 20));
  }

  log(`FINISH ${endpointPath}: fetched total of ${allRecords.length} records.`);
  fs.writeFileSync(
    path.join(DATA_DIR, fileName),
    JSON.stringify({ first_response_meta: firstResponse, total_records_fetched: allRecords.length, records: allRecords }, null, 2)
  );
  return allRecords;
}

async function main() {
  await fetchHealth();
  const token = await fetchAuth();
  if (!token) {
    log('Failed to get token!');
    return;
  }

  await fetchAllPages('/v1/listings', 'listings.json', token);
  await fetchAllPages('/v1/rentals', 'rentals.json', token);
  await fetchAllPages('/v1/projects', 'projects.json', token);

  log('All datasets successfully downloaded!');
}

main().catch(console.error);
